pkg/config/cfgvars.go

```diff
@@ -0,0 +1,303 @@
+		KubeletAuthConfigPath:      filepath.Join(dataDir, "kubelet.conf"),
+		KubeletVolumePluginDir:     constant.KubeletVolumePluginDir,
+		ManifestsDir:               filepath.Join(dataDir, "manifests"),
+		RunDir:                     runDir,
+		KonnectivityKubeConfigPath: filepath.Join(certDir, "konnectivity.conf"),
+		RuntimeConfigPath:          filepath.Join(runDir, "k0s.yaml"),
+		StatusSocketPath:           filepath.Join(runDir, "status.sock"),
+		StartupConfigPath:          constant.K0sConfigPathDefault,
+
+		// Helm Config
+		HelmHome:             helmHome,
+		HelmRepositoryCache:  filepath.Join(helmHome, "cache"),
+		HelmRepositoryConfig: filepath.Join(helmHome, "repositories.yaml"),
+
+		origin: CfgVarsOriginDefault,
+	}
+
+	if cmd != nil {
+		WithCommand(cmd)(vars)
+	}
+
+	return vars, nil
+}
+
+func (c *CfgVars) Cleanup() error {
+	if c.origin == CfgVarsOriginDefault && file.Exists(c.RuntimeConfigPath) {
+		return os.Remove(c.RuntimeConfigPath)
+	}
+	return nil
+}
+
+func (c *CfgVars) defaultStorageSpec() *v1beta1.StorageSpec {
+	if c.DefaultStorageType == v1beta1.KineStorageType {
+		return &v1beta1.StorageSpec{
+			Type: v1beta1.KineStorageType,
+			Kine: v1beta1.DefaultKineConfig(c.DataDir),
+		}
+	}
+
+	return v1beta1.DefaultStorageSpec()
+}
+
+func (c *CfgVars) NodeConfig() (*v1beta1.ClusterConfig, error) {
+	if c.nodeConfig != nil {
+		return c.nodeConfig, nil
+	}
+
+	if c.origin == CfgVarsOriginRuntime {
+		return nil, fmt.Errorf("runtime config is not available")
+	}
+
+	if c.StartupConfigPath == "" {
+		return nil, fmt.Errorf("config path is not set")
+	}
+
+	var nodeConfig *v1beta1.ClusterConfig
+
+	cfgContent, err := os.ReadFile(c.StartupConfigPath)
+	if errors.Is(err, os.ErrNotExist) && c.StartupConfigPath == constant.K0sConfigPathDefault {
+		nodeConfig = v1beta1.DefaultClusterConfig(c.defaultStorageSpec())
+	} else if err == nil {
+		cfg, err := v1beta1.ConfigFromString(string(cfgContent), c.defaultStorageSpec())
+		if err != nil {
+			return nil, err
+		}
+		nodeConfig = cfg
+	} else {
+		return nil, err
+	}
+
+	if nodeConfig.Spec.Storage.Type == v1beta1.KineStorageType && nodeConfig.Spec.Storage.Kine == nil {
+		nodeConfig.Spec.Storage.Kine = v1beta1.DefaultKineConfig(c.DataDir)
+	}
+
+	c.nodeConfig = nodeConfig
+
+	return nodeConfig, nil
+}
+
+func (c *CfgVars) ClusterConfig(ctx context.Context, kubeClientFactory kubeutil.ClientFactoryInterface) (*v1beta1.ClusterConfig, error) {
+	if !c.EnableDynamicConfig {
+		return c.NodeConfig()
+	}
+
+	var apiConfig *v1beta1.ClusterConfig
+
+	client, err := kubeClientFactory.GetConfigClient()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := retry.Do(
+		func() (err error) {
+			ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+			defer cancel()
+			c, err := client.Get(ctx, constant.ClusterConfigObjectName, metav1.GetOptions{})
+			if err == nil {
+				apiConfig = c.GetClusterWideConfig()
+			}
+			return err
```

Comment:
Intuitively, I think it's more common to write it like this:

```suggestion
			if err != nil {
				return err
			}

			apiConfig = c.GetClusterWideConfig()
			return nil
```